### ISSUE TITLE
Information request: How to input hessian for custom multi-class objective function?

### ISSUE THREAD

**Issue by HotSauceTea** [microsoft/LightGBM#4956]
## Description
Hello,
I'm trying to use a custom objective functions for a multi-class classification problem, and I'm confused about the ```hess``` parameter required.

The documentation says hess should be "The value of the second order derivative (Hessian) of the loss with respect to the elements of preds for each sample point." https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/basic.py#L2958-L2960
So for a loss function f(z_1, z_2, ... z_K), I'd expect to input the matrix of second derivatives d^2f/dz_idz_j at each sample (flattened to 1d as required).

However, this would have length N*K^2 if there are N samples and K classes. The gradient should have length N*K, but I see that the lengths of the gradient and the hessian must be the same:

https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/basic.py#L3034
 
How should I input the Hessian so that LightGBM can use it? Is it only the diagonal entries d^2f/dz_i^2?

I've only been able to find one example of a multi-class custom loss function (https://github.com/Microsoft/LightGBM/issues/1644), and if I'm understanding it right, in that case it's softmax regression loss and the Hessian is a diagonal matrix, so it makes sense that all you'd need is the diagonal entries. My use case is different and the hessian has off-diagonal entries.

Thanks for you help!

**Comment by jmoralez** [comment]
Hi @HotSauceTea. Yes, the input would be the diagonal in that case because the loss is assumed to be computed for a single sample, i.e. the loss function should only involve the target and the prediction for one sample.

For example, when using the regression objective, the loss function is `loss(y, pred) = 1/2 * (pred - y)^2`, so then the gradient is `dloss/dpred = pred - y` and the hessian `d^2loss/dpred^2 = 1`.

I don't think it's possible to have the loss depend on other samples, feel free to comment if that's what you're looking for.
